frame/support/procedural/src/benchmark.rs

```diff
@@ -705,36 +786,97 @@ fn expand_benchmark(
 				qself: None,
 				path: Path { leading_colon: None, segments: punct },
 			});
-
+			let pre_call = quote! {
+				let __call = Call::<#type_use_generics>::#expr_call;
+				let __benchmarked_call_encoded = #codec::Encode::encode(&__call);
+			};
+			let post_call = quote! {
+				let __call_decoded = <Call<#type_use_generics> as #codec::Decode>
+					::decode(&mut &__benchmarked_call_encoded[..])
+					.expect("call is encoded above, encoding must be correct");
+				let __origin = #origin.into();
+				<Call<#type_use_generics> as #traits::UnfilteredDispatchable>::dispatch_bypass_filter(
+					__call_decoded,
+					__origin,
+				)
+			};
 			(
-				// (pre_call, post_call):
-				quote! {
-					let __call = Call::<#type_use_generics>::#expr_call;
-					let __benchmarked_call_encoded = #codec::Encode::encode(&__call);
-				},
+				// (pre_call, post_call, fn_call_body):
+				pre_call.clone(),
+				quote!(#post_call?;),
 				quote! {
-					let __call_decoded = <Call<#type_use_generics> as #codec::Decode>
-						::decode(&mut &__benchmarked_call_encoded[..])
-						.expect("call is encoded above, encoding must be correct");
-					let __origin = #origin.into();
-					<Call<#type_use_generics> as #traits::UnfilteredDispatchable>::dispatch_bypass_filter(
-						__call_decoded,
-						__origin,
-					)?;
+					#pre_call
+					#post_call.unwrap();
 				},
 			)
 		},
-		BenchmarkCallDef::Block { block, attr_span: _ } => (quote!(), quote!(#block)),
+		BenchmarkCallDef::Block { block, attr_span: _ } =>
+			(quote!(), quote!(#block), quote!(#block)),
+	};
+
+	let vis = benchmark_def.fn_vis;
+
+	// remove #[benchmark] attribute
```

Comment:
Well the function def we generate in the expansion _is real_, it's just the visibility only applies to it, not to any of the other stuff we generate, which I think is reasonable. The alternative would be ignoring it (which would be confusing to the programmer when they try to set it to be pub or not pub and the generated functiond doesn't track with that), or restricting it to be a certain thing (pub or non pub), which I think would be needlessly restrictive.

Reply:
Hm my comment here was about attribute macros on the benchmarking function. Like `should_panic` or whatever would not make sense.

Reply:
Oh right, sorry. I would make the same argument, just about attributes:

I would lean towards just letting all attributes through here. Sure in some cases it might not make sense, but I think it would be more confusing if the programmer tries to use an attribute and it is ignored or was specifically blocked.

Alternatively, we could issue a compiler error for any non-doc attributes

Reply:
>Alternatively, we could issue a compiler error for any non-doc attributes.

That is what i mean.